src/python/WMCore/Services/DBS/DBS3Reader.py

```diff
@@ -950,20 +949,48 @@ def insertFileParents(self, childBlockName, childParentsIDPairs):
         """
         return self.dbs.insertFileParents({"block_name": childBlockName, "child_parent_id_list": childParentsIDPairs})
 
-    def findAndInsertMissingParentage(self, childBlockName, childLFNs=None, insertFlag=True):
+    def findAndInsertMissingParentage(self, childBlockName, parentData, insertFlag=True):
         """
         :param childBlockName: child block name
-        :param childLFNs: list of child lfns if it is not specified, all the file in the block will be used,
-               if specified, dbs validate child lfns from the childBlockName
+        :param parentData: a dictionary with complete parent dataset file/run/lumi information
+        :param insertFlag: boolean to allow parentage insertion into DBS or not
         :return: number of file parents pair inserted
         """
-        childLFNs = childLFNs or []
-        fileParents = self.dbs.listFileParentsByLumi(block_name=childBlockName, logical_file_name=childLFNs)
-        childParentsIDPairs = fileParents[0]["child_parent_id_list"]
+        # in the format of: {'fileid': [[run_num1, lumi1], [run_num1, lumi2], etc]
+        # e.g. {'554307997': [[1, 557179], [1, 557178], [1, 557181],
+        childBlockData = self.dbs.listBlockTrio(block_name=childBlockName)
+
+        # runs the actual mapping logic, like {"child_id": ["parent_id", "parent_id2", ...], etc
+        mapChildParent = {}
+        # there should be only 1 item, but we better be safe
+        for item in childBlockData:
```

Comment:
It looks like the complexity of this step to be O(n^3). I know it is impossible to avoid it here..., my question is more about did you measure if this is actually a significant time consuming step and is this time comparable to the time used for waiting for DBS and so on....

Reply:
From what I can see, childBlockData will always have one element only (reason why I put that comment in the line above). So practical complexity will be smaller than that.
I do not have heavy data in DBS pre-prod, so I cannot test the performance of this (besides the ad-hoc script created when I was proposing possible solutions/algorithm for this).